pkg/csi/service/wcp/controller.go

```diff
@@ -1021,6 +1021,15 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 		}
 		if !isVADeleted {
 			log.Debugf("controllerUnpublishVolumeInternal: VA : %v", spew.Sdump(volumeAttachment))
+			// Check if the NodeName in the VA is same as the NodeName from the ControllerUnPublishVolume.
+			// This is needed to return a successful ControllerUnPublishVolume response when a pod is
+			// rescheduled onto another worker node(ESX host). This is a no-op for CNS-CSI in Supervisor.
+			if volumeAttachment.Spec.NodeName != req.NodeId {
```

Comment:
`volumeAttachment.Spec.NodeName` is hashed as part of `VolumeAttachment.Name` computed [here](https://github.com/chethanv28/vsphere-csi-driver/blob/cab5ff7a67a6c33057491b653a7c8b4380b3b197/pkg/csi/service/common/commonco/k8sorchestrator/k8sorchestrator.go#L1356). How is it possible that this `if` is ever true?

Reply:
@jsafrane We observed this in one of our environments, where there were two different volumeAttachment objects for the same PV. One(say VA1) indicated the correct node for the volume and the other one had deletion timestamp set and a different node(say VA2). They had different nodeIds, but same VM in the `Status.AttachmentMetadata` which is populated in publishContext when returning from attach volume call.

While processing ControllerUnpublishVolume for VA2, we were checking if the volume is really detached from the VM before returning success. But since VA1 also had the same VM in it's status, we were failing this call and it kept retrying.

I'm not sure how we landed up in this state. It may be a bug during CSI driver upgrade, or a bug in external-attacher. But to remediate this scenario, we made this change.